src/Processors/Transforms/MergeJoinTransform.cpp

```diff
@@ -0,0 +1,865 @@
+    PaddedPODArray<UInt64> idx_map[2];
+    size_t prev_pos[] = {current_left.getRow(), current_right.getRow()};
+
+    dispatchKind<AnyJoinImpl>(kind, *cursors[0], *cursors[1], idx_map[0], idx_map[1], any_join_state);
+
+    assert(idx_map[0].empty() || idx_map[1].empty() || idx_map[0].size() == idx_map[1].size());
+    size_t num_result_rows = std::max(idx_map[0].size(), idx_map[1].size());
+
+    /// build result block from indices
+    Chunk result;
+    for (size_t source_num = 0; source_num < 2; ++source_num)
+    {
+        /// empty map means identity mapping
+        if (!idx_map[source_num].empty())
+        {
+            for (const auto & col : cursors[source_num]->getCurrent().getColumns())
+            {
+                result.addColumn(indexColumn(col, idx_map[source_num]));
+            }
+        }
+        else
+        {
+            for (const auto & col : cursors[source_num]->getCurrent().getColumns())
+            {
+                result.addColumn(col->cut(prev_pos[source_num], num_result_rows));
+            }
+        }
+    }
+    return Status(std::move(result));
+}
+
+/// if `source_num == 0` get data from left cursor and fill defaults at right
+/// otherwise - vice versa
+Chunk MergeJoinAlgorithm::createBlockWithDefaults(size_t source_num, size_t start, size_t num_rows) const
+{
+    ColumnRawPtrs cols;
+    {
+        const auto & columns_left = source_num == 0 ? cursors[0]->getCurrent().getColumns() : sample_chunks[0].getColumns();
+        const auto & columns_right = source_num == 1 ? cursors[1]->getCurrent().getColumns() : sample_chunks[1].getColumns();
+
+        for (size_t i = 0; i < columns_left.size(); ++i)
+        {
+            if (auto it = left_to_right_key_remap.find(i); source_num == 0 || it == left_to_right_key_remap.end())
+            {
+                cols.push_back(columns_left[i].get());
+            }
+            else
+            {
+                cols.push_back(columns_right[it->second].get());
+            }
+        }
+
+        for (const auto & col : columns_right)
+        {
+            cols.push_back(col.get());
+        }
+    }
+
+    Chunk result_chunk;
+    copyColumnsResized(cols, start, num_rows, result_chunk);
+    return result_chunk;
+}
+
+/// This function also flushes cursor
+Chunk MergeJoinAlgorithm::createBlockWithDefaults(size_t source_num)
+{
+    Chunk result_chunk = createBlockWithDefaults(source_num, cursors[source_num]->cursor.getRow(), cursors[source_num]->cursor.rowsLeft());
+    cursors[source_num]->detach();
+    return result_chunk;
+}
+
+IMergingAlgorithm::Status MergeJoinAlgorithm::merge()
+{
+    auto kind = table_join->getTableJoin().kind();
+
+    if (!cursors[0]->cursor.isValid() && !cursors[0]->fullyCompleted())
+        return Status(0);
+
+    if (!cursors[1]->cursor.isValid() && !cursors[1]->fullyCompleted())
+        return Status(1);
+
+    if (auto result = handleAllJoinState())
```

Comment:
Just I am curious, why call `handleAllJoinState()` method here instead of calling it in `allJoin()`, and calling `handleAnyJoinState` is located in `anyJoin()` method.

Reply:
To be honest I don't remember exactly. Perhaps `handle*State` functions rely on the code below (`/// check if blocks are not intersecting at all`) in a different ways. Or it's also possible that there's no particular reason, need to dive into the code to understand.